Add support for camera on i.MX RT11XX

ngphibang opened the pull request:
This PR adds support to enable camera on i.MX RT11XX series. It adds

- ov5640 camera driver
- MIPI CSI-2 Rx driver
- Modify CSI driver to support both RT11XX and RT10XX
- Improve the video capture sample to display frames on LCD panel instead of just discarding them. For this to work, [another PR](https://github.com/zephyrproject-rtos/zephyr/pull/69250) should be merged first.

ngphibang committed:
soc: nxp: imxrt11xx: Enable camera master clock

Enable camera master clock for i.MX RT11XX series

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Change sensor dev to source dev

The CSI can connect to either a camera sensor (as on RT10xx) or a MIPI
CSI-2 receiver (as on rt 11xx). To be generic, change the naming from
sensor dev to source dev.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Drop pixel format config

The pixel format config is not used in the low level driver. Drop it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts: arm: nxp: Add devicetree node for MIPI CSI-2 Rx

Add a node for MIPI CSI-2 Rx in i.MX RT11xx devicetree and connect it to
the CSI node.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
soc: nxp: imxrt11xx: Configure clocks and muxing for MIPI CSI-2 Rx

Enable clocks for MIPI CSI-2 Rx and configure Video Mux to connect it to
CSI.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
samples: video: capture: Use a chosen node for camera

In order to be generic, use a chosen node for camera so that the sample
is not specific to NXP SoCs. Also, always favorite a real video device
unless it is unavailable.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
boards: nxp: mixmxrt1064_evk: Drop camera sensor related stuffs

The mt9m114 camera module is not always attached to this board. Camera
stuffs should go into a seperate overlay / shield. Also, the CSI should
be enabled only if there is a camera sensor attached to it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
boards: nxp: mixmxrt1170_evk: Separate i2c6 from csi pin muxing

Pin muxing for i2c6 and csi should be separated as i2c is initialized
before csi. Otherwise, i2c6 bus device will not be ready when the
camera sensor on this bus initializes.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Add support for i.MX RT11XX

On i.MX RT11XX which has MIPI CSI-2 Rx, image data from the camera sensor
after passing through the camera pipeline (MIPI CSI-2 Rx --> Video Mux -->
CSI) will be implicitly converted to a 32-bits pixel formats. For example,
an input in RGB565 / YUYV (2-bytes format) will become an XRGB32 / XYUV32
(4-bytes format), respectively. Make changes to support this.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
samples: video: capture: Display captured frames on screen

Improve the sample application by displaying the captured frames instead
of discarding them.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
samples: video: capture: Run clang-format

Run clang-format on the file before making any changes

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Add NXP MIPI CSI-2 Rx driver

Add driver to support NXP MIPI CSI-2 Rx which is a MIPI CSI-2 receiver
connecting a camera sensor to the NXP CSI. This IP is present in the
i.MX RT11XX series.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts: bindings: video: Add bindings for NXP MIPI CSI-2 Rx

Add bindings for NXP MIPI CSI-2 Rx which is a MIPI CSI-2 receiver
connecting a camera sensor to the NXP CSI.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Add ov5640 camera driver

Add driver to support ov5640 camera sensor

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts: bindings: Add bindings for ov5640 camera

Add bindings for ov5640 camera sensor

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Run clang format

Run clang format before making any changes

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
samples: video: capture: Increase video buffer size for i.MX RT11XX

The default video buffer pool max size is not enough to afford for 720p
resolution and XRGB32 pixel format which is the default resolution and
format for the camera pipeline on i.MX RT11XX. Increase it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
boards: shields: Add shield for ov5640 camera module

Add a shield for the ov5640 camera module. This camera module is
supported on i.MX RT11xx SoC family.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
boards: shields: Add shield for mt9m114 camera module

Add a shield for the mt9m114 camera module. This camera module is
supported on i.MX RT10xx SoC family.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
tests: drivers: build_all: video: Add entry for ov5640

Add entry for ov5640 and add mimxrt1170_evk to the test platforms

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
Hi @danieldegrasse , @decsny , @DerekSnell, could you please advise on the naming of the MIPI CSI-2 Rx driver ?

Currently, I named it "mipi_csi2rx_nxp" because I remembered @DerekSnell has said that :

> We should not use imx in the name.  We reuse our IP everywhere across SOC families.  So we should not include SOC family names in the drivers or bindings.

@decsny also proposed "CONFIG_VIDEO_MCUX_MIPI_CSI2RX". I think it is correct too. But frankly, I am a bit confused between "mcux" and "imx" (imx seems covers both mcu and mpu ?). So, I hesitated too much. 

Is it ok for you if we name them as below ?

- CONFIG_VIDEO_MCUX_MIPI_CSI2RX
- video_mcux_mipi_csi2rx.c
- Kconfig.mcux_mipi_csi2rx

Thanks

decsny commented:
> But frankly, I am a bit confused between "mcux" and "imx" (imx seems covers both mcu and mpu ?).

MCUX refers to the MCUX-SDK. Which can theoretically have any kind of device in it, it's just a software ecosystem. IMX refers to the IMX MPU family. For the IMX MCU (RT family), they are called by IMXRT. (see soc/nxp folder).

DerekSnell commented:
Hi @ngphibang ,
Thank you for confirming these names.  
> But frankly, I am a bit confused between "mcux" and "imx" (imx seems covers both mcu and mpu ?).

I understand.  To add to Declan's response: imx, or i.MX, is a marketing brand name for products.  I do not want us using product family names when naming Zephyr drivers or bindings, even though NXP has done this in the past.  Hardware IP is re-used across different product families, so drivers/bindings should not restrict to specific families.

"mcux" is short for MCUXpresso, which is our enablement ecosystem, including the MCUXpresso SDK.  Our Zephyr naming uses mcux to indicate the driver or feature utilizes the SDK APIs or files in our HAL.  

In your case, naming with mcux makes sense.  And I appreciate you avoiding imx.

> Is it ok for you if we name them as below ?
> 
> * CONFIG_VIDEO_MCUX_MIPI_CSI2RX
> * video_mcux_mipi_csi2rx.c
> * nxp,mipi-csi2rx.yaml
> * Kconfig.mcux_mipi_csi2rx

Yes, those names look great to me.  Thank you

ngphibang committed:
boards: nxp: mimxrt1170_evk: Enable video feature

Enable video feature on cm7 evk and evkb which are tested with ov5640
camera.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
samples: video: capture: Update documents and tests

Update documents and tests to support the i.MX RT1170 EVK, the video
test pattern generator and the new camera shields.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
MIPI CSI-2 Rx naming is changed with the new update.

ngphibang commented:
Hi @loicpoulain , @danieldegrasse,
Could you help to review this PR (this has been here for a while :-) ) ?
All checks are passed (the Bsim test is failed but not related to the PR).
Thanks !

ngphibang commented:
hal_nxp PR is merged.  I updated the new SHA revision

ngphibang commented:
@loicpoulain : Hi Loic, would you help to review this one :-) ?
Thanks !

loicpoulain commented:
> @loicpoulain : Hi Loic, would you help to review this one :-) ?
> 
 Yes, it's a large PR, need some time to review that properly.

ngphibang commented:
> > @loicpoulain : Hi Loic, would you help to review this one :-) ?
> 
> Yes, it's a large PR, need some time to review that properly.

Yes, it does :-). I tried to put independent things into separated PRs. But for these, I think it is better to put them together as they are dependent and all for adding support camera on RT11xx. In fact, the main parts are :

- ov5640 driver
- mipi csi-2 driver
- [drivers: video: csi: Add support for i.MX RT11XX](https://github.com/zephyrproject-rtos/zephyr/pull/69810/commits/8778665546fe74f76934f39c6bfd8d6b17d8116f)
- video capture sample changes to support display.

Others are minor changes, documents, tests that must come with the main parts.
Thanks.

dleach02 commented on drivers/video/Kconfig.mcux_mipi_csi2rx:
```diff
@@ -0,0 +1,10 @@
+# NXP MIPI CSI-2 Rx driver configuration option
+
+# Copyright (c) NXP 2024
```
Guidance on NXP copyright is no '(c)'

ngphibang replied:
yes, done

dleach02 commented on drivers/video/Kconfig.ov5640:
```diff
@@ -0,0 +1,12 @@
+# OV5640
+
+# Copyright (c) 2024 NXP
```
no '(c)'

ngphibang replied:
yes, done

dleach02 commented on drivers/video/ov5640.c:
```diff
@@ -0,0 +1,670 @@
+/*
+ * Copyright (c) 2024, NXP
```
no '(c)'

ngphibang replied:
yes, done

danieldegrasse commented on drivers/video/ov5640.c:
```diff
@@ -0,0 +1,670 @@
+	{0x3008, 0x42},
+
+	/* System setting. */
+	{0x3103, 0x13},
```
It looks like there is at least *limited* documentation on these registers present online: https://cdn.sparkfun.com/datasheets/Sensors/LightImaging/OV5640_datasheet.pdf. Can we please add `#define` definitions for these registers where possible?

danieldegrasse commented on drivers/video/ov5640.c:
```diff
@@ -0,0 +1,670 @@
+	return ov5640_init(dev);
+}
+
+DEVICE_DT_INST_DEFINE(0, &ov5640_init_0, NULL, &ov5640_data_0, &ov5640_cfg_0, POST_KERNEL,
```
Why not make this driver multi-instance? For SOC IP blocks (where only one instance exists on silicon) I can see the point of keeping a driver single instance, but a user can easily purchase and potentially connect two cameras for their application, right?

ngphibang replied:
Hi @loicpoulain, do you know if there is a specific reason that the mt9m114 and other camera drivers were explicitly made as unique instance ?

ngphibang replied:
Don't get the answer from @loicpoulain, turned the driver into multi-instance though

danieldegrasse commented on dts/bindings/video/ovti,ov5640.yaml:
Commit with the OV5640 DT binding should come prior to the commit with the camera driver

danieldegrasse replied:
This is a good point- hadn't considered this. I guess it does not matter either way in that case

ngphibang replied:
Although it doesn't matter but I changed the order anyways.

danieldegrasse commented on dts/bindings/video/ovti,ov5640.yaml:
```diff
@@ -0,0 +1,19 @@
+    description: |
+      The PWDN pin is asserted to disable the sensor. The sensor
+      receives this as an active-high signal.
+include: i2c-device.yaml
```
nit- include the binding at the top of the file

ngphibang replied:
done

danieldegrasse commented on tests/drivers/build_all/video/testcase.yaml:
```diff
@@ -13,4 +13,6 @@ tests:
       - gpio
       - i2c
   drivers.video.mcux_csi.build:
```
We shouldn't be adding the RT1170 to the CSI build test until after we add support to the driver, right? So this change shouldn't be the commit adding the ov5640 to the build all test

ngphibang replied:
done

danieldegrasse commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -0,0 +1,215 @@
+		},
+	};
+
+	csi2rxConfig.laneNum = DEFAULT_MIPI_CSI_NUM_LANES;
```
We should probably make this configurable with devicetree, right? For MIPI-DSI drivers we have a discrete interface that supports attaching and detaching (which we might want for MIPI-CSI drivers in the future as well). For now though, there should still be a way to change this without modifying the source code IMO.

danieldegrasse replied:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/34665051/81796442-0113-457d-ba8e-3a9537233906)

It seems the MIPI CSI2 supports 1 or 2 data lanes. While I agree there is probably work to be done extending the video API with endpoints, why wouldn't we expose this right now as a configurable property? This feels like something a customer will run into pretty soon with this driver if we lock them to using 2 lanes when the IP supports 1 data lane

ngphibang replied:
Sorry, it is miswritten in my previous comment. I mean the ov5640 driver does not support 1 data lane (not the MIPI CSI2Rx). There is currently no camera driver that supports 1 data lanes. So, I think we should not expose this in the devicetree (even as a temporary solution) as it is never tested on any hardware.

In the future, when there is a camera driver supporting it, we could do so as your suggestion (again as a temporary solution). But **the data lanes need to be declared in both MIPI CSI2RX node and the camera sensor node** (and not only the number of lanes, the order of the lanes does matter also).

Another reason, 1-data lane camera sensor is rare as it is not efficient as 2 or 4 data lanes.

danieldegrasse commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -0,0 +1,215 @@
+LOG_MODULE_REGISTER(mipi_csi);
+
+#define DEFAULT_MIPI_CSI_NUM_LANES 2
+#define DEFAULT_CAMERA_FRAME_RATE  30
```
Should we make this configurable with devicetree?

danieldegrasse replied:
> There are a few "parameters" that can be declared in the devicetree, for example
>
>   -  the camera orientation / rotation
 >  -  data-lanes / clock-lanes

Isn't one of the parameters mentioned here to move to devicetree the number of data lanes? It seems like that is an excellent candidate for devicetree. As far as the framerate, I do agree this might not be best for devicetree. *Maybe* Kconfig? I'm not sure where best to put this.

ngphibang replied:
The _ov5640 driver does not support framerate changing_, (even the much more complete ov5640 driver in Linux only supports framerate changing in parallel mode, not csi). That's the main point.

Even it supports framerate changing. As being said, this needs to be done in a _negotiation at runtime_, between application, camera sensor, mipi csi2, csi (similar to set_format)... A simple expose in Kconfig or devicetree will not work, I think.

danieldegrasse replied:
Ok- thanks for the clarification :) I wasn't aware that there would need to be runtime negotiation here- we can leave this value as is

danieldegrasse commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -0,0 +1,215 @@
+	}
+
+	/* MIPI DPHY power on and isolation off */
+	PGMC_BPC4->BPC_POWER_CTRL |= (PGMC_BPC_BPC_POWER_CTRL_PSW_ON_SOFT_MASK |
```
This seems like it is probably specific to the RT11xx. We should either:
- guard this with `CONFIG_SOC_SERIES_IMXRT11XX`
- add a function like [imxrt_pre_init_display_interface](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/imxrt/imxrt5xx/cm33/soc.h#L80) for the MIPI CSI, and implement it for the RT11XX

ngphibang replied:
MIPI CSI-2 Rx is an IP specific to RT11xx (RT10xx doesn't have this IP). So, I think it's ok to set this register here, in the driver. It's not the same case as the display (eLCDIF is common IP for both 10xx and 11xx)

danieldegrasse replied:
> MIPI CSI-2 Rx is an IP specific to RT11xx (RT10xx doesn't have this IP). So, I think it's ok to set this register here, in the driver. It's not the same case as the display (eLCDIF is common IP for both 10xx and 11xx)

It is specific *now*. There is no promise that future SOCs will not integrate this IP, and at that point code like this within the driver layer becomes an issue.

ngphibang replied:
Okay, I will guard it with "CONFIG_SOC_SERIES_IMXRT11XX".

danieldegrasse commented on dts/bindings/video/nxp,mipi-csi2rx.yaml:
Same comment on DT bindings- the binding needs to come before the driver in the commit series

ngphibang replied:
Although it doesn't matter but I changed the order anyways.

danieldegrasse commented on soc/nxp/imxrt/imxrt11xx/soc.c:
```diff
@@ -456,6 +456,23 @@ static ALWAYS_INLINE void clock_init(void)
+	CLOCK_SetRootClock(kCLOCK_Root_Csi2_Esc, &rootCfg);
+
+	rootCfg.mux = kCLOCK_CSI2_UI_ClockRoot_MuxSysPll3Out;
+	CLOCK_SetRootClock(kCLOCK_Root_Csi2_Ui, &rootCfg);
```
This is probably where we should put the code writing to the PGMC register, if possible

danieldegrasse replied:
Sure- that is a good idea, and I'd support (at a minimum) making things like the video mux setting configurable in devicetree somehow. A simple video driver that reads the setting from the DT node could work well here. I just want to avoid SOC specific code in the driver for an IP block whenever possible, even if the IP is currently only present on one SOC we have enabled

ngphibang replied:
It's done [here](https://github.com/zephyrproject-rtos/zephyr/pull/71859).

danieldegrasse commented on soc/nxp/imxrt/imxrt11xx/Kconfig:
```diff
@@ -45,6 +45,7 @@ config SOC_SERIES_IMXRT11XX
 	select HAS_MCUX_XBARA
 	select HAS_SWO
 	select HAS_PM
+	select HAS_MCUX_CSI
```
Why do we need to select this KConfig? These Kconfigs have largely been replaced with `DT_HAS_XXX` Kconfigs, and are legacy from before that feature existed

danieldegrasse commented on drivers/video/video_mcux_csi.c:
```diff
@@ -126,7 +126,6 @@ static int video_mcux_csi_set_fmt(const struct device *dev, enum video_endpoint_
 		return -EINVAL;
 	}
 
-	data->pixelformat = fmt->pixelformat;
```
At least within this commit, isn't `video_mcux_csi_get_fmt` still using `data->pixelformat`? Might be missing something here.

danieldegrasse commented on dts/bindings/video/nxp,imx-csi.yaml:
This binding change needs to be moved to before the CSI camera definition is added to the RT11xx devicetree, in order to keep these commits bisectable (since you use `sdev` in that DTS entry)

danieldegrasse commented on drivers/video/video_mcux_csi.c:
```diff
@@ -114,23 +117,62 @@ static void __frame_done_cb(CSI_Type *base, csi_handle_t *handle, status_t statu
 }
 
+#if defined(CONFIG_VIDEO_MCUX_MIPI_CSI2RX)
+K_HEAP_DEFINE(csi_heap, 1000);
```
Would it not be possible to statically allocate the space for the CSI formats within the device data struct, and initialize them in `video_mcux_csi_set_fmt`? It doesn't seem like we need to use a heap for this single function call, versus statically allocated memory (since allocation is generally not desirable in some embedded applications)

ngphibang replied:
No, I think it's not possible. The formats caps come frome the camera sensor. We have no information about it until it is propagated from the sensor (via get_caps()).  Since it (video_format_cap) is a const struct pointer defined by the API, I have to declare a new struct to change it (unless **_we agree to de-const this struct_** in the video API but I don't think it is a good way).

danieldegrasse commented on boards/shields/wuxi_ov5640/wuxi_ov5640.overlay:
```diff
@@ -0,0 +1,47 @@
+	};
+};
+
+&lpi2c6 {
```
In the past when defining shields for a NXP specific interface (which I believe this is?) we have named the interface something like `nxp_mipi_i2c` within the board devicetree- as this overlay is written, it will *only* work on the RT1170 EVK or any other EVK that happens to use LPI2C6. See here for an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/shields/rk055hdmipi4ma0/rk055hdmipi4ma0.overlay#L27

ngphibang replied:
In fact, lpi2c6 is used across all RT11XX family and I suppose this shield will work only for RT11XX family.
But yes, we should be generic. I will use some alias so that the shield is as generic as possible.

danieldegrasse commented on boards/shields/huatian_mt9m114/huatian_mt9m114.overlay:
```diff
@@ -0,0 +1,35 @@
+	};
+};
+
+&lpi2c1 {
```
Same comment as other shield overlay- we can't use the node labels that are specific to the board here, we need to define more generic labels that could be reused on other boards besides the RT1060 EVK

ngphibang replied:
If I remember well lpi2c1 is used across EVK RT1060, RT1064, etc.  for camera ?
But yes, we should be generic. I will use some alias so that the shield is as generic as possible.

danieldegrasse replied:
Yes, it very well may be- however, shields should _theoretically_ be able to work on a customer board, or another vendor's board provided they expose the same interfaces

ngphibang replied:
done

danieldegrasse commented on samples/subsys/video/capture/src/main.c:
Thanks for the work improving this sample- I think this should go into another PR though. It is a welcome addition to be clear, I just think it is functionally separate from the SOC/camera module support being added in this PR and should not be dependent on it.

ngphibang replied:
The code is somehow independent but the documents is dependent (on the shield definitions) and will not build if separated from this PR. Some changes on the sample only make sense if come with the support of camera on RT11xx so I put them here so that reviewers can test them. But yes, it is better to make another PR for these (and wait for this PR to be merged)

danieldegrasse replied:
Ok, that works for me. Do you think the other PR could enable the sample using the RT10xx, since the camera and display support is in tree? If not, I am totally ok with having the video sample improvements be dependent on this PR

ngphibang replied:
The changes in the sample is **dependant** on the shield changes (and other things) of this PR. Moreover, this PR **has been here for quite a while** . For both reasons, I wish to let them here so that we don't need to make a dependent PR and wait anymore for it to be merged.

danieldegrasse commented on samples/subsys/video/capture/README.rst:
```diff
@@ -7,9 +7,8 @@
-This sample application uses the :ref:`Video API <video_api>` to retrieve video frames from the
-video capture device, writes a frame count message to the console, and then
-discards the video frame data.
+This sample application uses the :ref:`Video API <video_api>` to capture frames from
```
These changes should be in the commit updating the Video capture sample

ngphibang replied:
Yes, thanks.

ngphibang replied:
done

danieldegrasse commented on boards/nxp/mimxrt1170_evk/doc/index.rst:
```diff
@@ -168,6 +168,9 @@ RT1170 EVKB (`mimxrt1170_evk@B//cm7/cm4`)
 +-----------+------------+-------------------------------------+-----------------+-----------------+
 | PIT       | on-chip    | pit                                 | Supported (M7)  | Supported (M7)  |
 +-----------+------------+-------------------------------------+-----------------+-----------------+
+| VIDEO     | on-chip    | CSI; MIPI CSI-2 Rx. Tested with     | Supported (M7)  | Supported (M7)  |
```
This can probably be in the commit where we add video support to the YAML for the RT1170 EVK

ngphibang replied:
Okay, I will merge these two commits.

danieldegrasse commented on west.yml:
```diff
@@ -193,7 +193,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: 3c4f74b405ca174fe3eb5f3b74c355376ee9a69f
+      revision: db6806f8a8207dd2e219aedd9ed9174f2cde8169
```
HAL NXP update needs to be prior to the changes to enable the CSI driver on the RT11xx to preserve bisectability

ngphibang replied:
yes - I think you mean MIPI CSI-2 driver ? as hal_update is for MIPI CSI-2

danieldegrasse replied:
Yes please, thanks!

ngphibang reviewed (commented):
Hi @danieldegrasse , 
Firstly, many thanks for your time reviewing this !
I will try to address your comments in my possibility. However, there are some that need to be revisited though (and also one comment needs feedback from @loicpoulain) before I made the changes.

dleach02 commented:
@ngphibang, while addressing the comments, don't forget to rebase to clear the merge conflict.

ngphibang committed:
modules: mcux: Drop HAS_MCUX_CSI config

The HAS_MCUX_CSI (as well as all the HAS_MCUX_XXX) config was obsolete
and has been replaced by the DT_HAS_NXP_IMX_CSI_ENABLED (i.e.
DT_HAS_XXX_ENABLED). Drop it as well as all the dependencies on it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Fix get_fmt

The CSI should always propagate the get_fmt request down to the sensor.
It should not return a format by it own unless it contains a test
pattern generator inside it.

The problem with the legacy code is that this code snippet is never
reached if the sensor could return a format. In case the sensor failed
to return a format, these codes are reached but the returned format
makes no sense and its value was not initialized.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
@danieldegrasse I updated the PR which addressed all of your comments in my possibility.

@loicpoulain , @danieldegrasse , @dleach02 : **Really need your helps to make this PR moving forward** as it has been here for quite a while.

Thanks !

loicpoulain commented:
@ngphibang could you split this PR in different logical PRs (like one for misc fixes, one for new camera driver, etc...), it will be easier for you to have them reviewed and integrated.

ngphibang commented:
@loicpoulain Okay, I will do it. But the problem is the next PR will depend on all previous PRs and Zephyr requires that each commit / PR is bisectable and buildable.

Would you mind if I include all the commits of the previous PRs into another PR ? For example :

PR 1 
PR 2  = its commits + commits of PR 1
PR 3 = its commits + all PR2

ngphibang commented:
@loicpoulain The first 3 splitted PRs which are independent each others are pushed.
Waiting for your reviews and integration before continuing to push the next ones which depend on these.

mmahadevan108 commented on drivers/video/video_mcux_csi.c:
```diff
@@ -116,25 +117,62 @@ static void __frame_done_cb(CSI_Type *base, csi_handle_t *handle,
 		return -EINVAL;
 	}
 
-	data->pixelformat = fmt->pixelformat;
```
This can be deleted from `video_mcux_csi_data`

ngphibang replied:
Yes, it is addressed and merged.

mmahadevan108 commented on dts/arm/nxp/nxp_rt11xx.dtsi:
```diff
@@ -1230,3 +1230,7 @@
 	status = "disabled";
 };
+
+zephyr_cam_i2c: &lpi2c6 {};
```
This seems board specific and probably belongs in the board overlay file.

ngphibang replied:
yes, it is pointed out by Daniel and I addressed it in the split PR.

mmahadevan108 commented on boards/nxp/mimxrt1064_evk/mimxrt1064_evk.dts:
```diff
@@ -254,16 +242,8 @@ zephyr_udc0: &usb1 {
-		csi_ep_in: endpoint {
-			remote-endpoint = <&mt9m114_ep_out>;
-		};
-	};
```
Do we need to add the labels related to the new shield you have created for 1064?

ngphibang commented:
@mmahadevan108 : Thank you for your reviews ! But this PR is split into smaller one as requested by Loic. Could you review them instead ? I don't know if we can add a label "do not review" or "stale" or something like that on this PR ?

mmahadevan108 commented on samples/subsys/video/capture/README.rst:
```diff
@@ -7,9 +7,8 @@
 Description
 ***********
```
Has this been tested in 1064_evk after the changes to the sample? Do we need a boards overlay file for 1064_evk?

ngphibang replied:
yes, it's tested. The board overlay for 1064 evk is in the same PR.

danieldegrasse commented:
@ngphibang is this PR still needed? My understanding was that all the components of this PR had been split into smaller PRs. If we don't need this one open, let's go ahead and close it out to avoid additional review noise

ngphibang commented:
I thought that we should keep it as a reference and just close it when all the split PRs merged.
I close it now, sorry.

ngphibang commented:
Close to avoid additional review noise

danieldegrasse commented:
> I thought that we should keep it as a reference and just close it when all the split PRs merged.
I close it now, sorry.

No worries- just don't want anyone reviewing the enablement to get confused by this PR :)
